rfcs/0014-transition-frontier-persistence.md

```diff
@@ -0,0 +1,112 @@
+
+In order to calculate this hash correctly, we need to introduce a new concept to a diff, which is that of a diff mutant. Each diff represents some mutation to perform on the Transition Frontier, however not every diff will contain the enough information by itself to encapsulate the state of the data structure after the mutation occurs. For example, setting a balance on an account in two implementations of the data structure does not guarantee that the accounts in each a equal as there are other fields on the account besides that. This is where the concept of a diff mutant comes in. The mutant of a diff is the set of all modified values in the data structure after the diff has been applied. Using this, we can create a proper incremental diff which will truly ensure our data structures are in sync.
+
+These hashes will be Sha256 as there is no reason to use the Pedersen hashing algorithm we use in the rest of our code since none of this information needs to be snarked. The formula for calculating a new hash `h'` given an old hash `h` and a diff `diff` is as follows: `h' = sha256 h diff (Diff.mutant diff)`.
+
+### Diff Application and Incremental Hash Pseudo-code
+[detailed-design-diff-application-and-incremental-hash-pseudo-code]: #detailed-design-diff-application-and-incremental-hash-pseudo-code
+
+Here is some pseudo code which models the `Diff.t` type as a GADT where the parameter type is the type of the mutant the diff returns. This only models a few of the diff types and most likely needs to be updated as this was first drafted before the Transition Frontier Extension RFC landed.
+
+```ocaml
+module Diff = struct
+  module T = struct
+    type 'mutant t =
+      | AddAccount : Public_key.Compressed.t * Account.t -> (Path.t * Account.t) t
+      | SetHash : Path.t * Hash.t -> Hash.t t
+      | SetBalance : Public_key.Compressed.t * Balance.t -> Account.t t
+    [@@deriving hash, eq, sexp]
+  end
+
+  include T
+
+  module Hlist = Hlist.Make (T)
+
+  let map_mutant
+      (diff : 'mutant t)
+      (mutant : 'mutant) 
+      ~(add_account : Path.t -> Account.t -> 'result)
+      ~(set_hash : Hash.t -> 'result)
+      ~(set_balance : Account.t -> 'result)
+      : 'result
+  = fun diff mutant ~add_account ~set_hash ~set_balance ->
+    match diff, mutant with
+    | AddAccount _, (path, account) -> add_account path account
+    | SetHash _, hash -> set_hash hash
+    | SetBalance _, account -> set_balance account
+end
+
+let apply_diff (t : t) (diff : 'mutant Diff.t) : 'mutant =
```

Comment:
I tried implementing this. I got a weird compiling error with 'mutant. If you use locally abstract types like the follow below, you would not get a compiling error and it will understand you:

`let apply_diff (type mutant) (t : t) (diff : mutant Diff.t) :  `